File: 剑指 Offer 34. 二叉树中和为某一值的路径.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> ret;
    vector<int> path;
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        dfs(root,targetSum);
        return ret;
    }
    void dfs(TreeNode* root, int targetSum)
    {
        if(root == nullptr) return;
        path.push_back(root->val);
        targetSum -= root->val;
        if(targetSum == 0 && root->left == nullptr && root->right == nullptr)
        {
            ret.emplace_back(path);
        }
        dfs(root->left, targetSum);
        dfs(root->right, targetSum);
        path.pop_back();
    }
};

File: 剑指 Offer 34. 二叉树中和为某一值的路径.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    LinkedList<List<Integer>> ret = new LinkedList<List<Integer>>();
    Deque<Integer> path = new LinkedList<Integer>();

    public List<List<Integer>> pathSum(TreeNode root, int target) {
        dfs(root, target);
        return ret;
    }

    public void dfs(TreeNode node, int target) {
        if (node == null)
            return;
        path.offerLast(node.val);

        target -= node.val;
        if (node.left == null && node.right == null && target == 0) {
            ret.add(new LinkedList<Integer>(path));
        }
        dfs(node.left, target);
        dfs(node.right, target);
        path.pollLast();
    }
}